fungible_proxy/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::json_types::{ValidAccountId, U128};
+use near_sdk::serde::{Deserialize, Serialize};
+use near_sdk::serde_json::json;
+use near_sdk::{
+    env, near_bindgen, serde_json, AccountId, Balance, Gas, Promise
+};
+near_sdk::setup_alloc!();
+
+const NO_DEPOSIT: Balance = 0;
+const YOCTO_DEPOSIT: Balance = 1; // Fungible token transfers require a deposit of exactly 1 yoctoNEAR
+const MIN_GAS: Gas = 150_000_000_000_000;
+const BASIC_GAS: Gas = 10_000_000_000_000;
+
+/// Helper struct containing arguments supplied by the caller
+///
+/// - `amount`: in `currency` with 2 decimals (eg. 1000 is 10.00)
+/// - `fee_address`: `fee_amount` in `currency` of payment token will be paid to this address
+/// - `fee_amount`: in `currency`
+/// - `payment_reference`: used for indexing and matching the payment with a request
+/// - `to`: `amount` in `currency` of payment token will be paid to this address
+#[derive(Serialize, Deserialize)]
+pub struct PaymentArgs {
+    pub fee_address: ValidAccountId,
+    pub fee_amount: U128,
+    pub payment_reference: String,
+    pub to: ValidAccountId,
+}
+
+impl Into<PaymentArgs> for String {
+    fn into(self) -> PaymentArgs {
+        serde_json::from_str(&self).expect("Incorrect msg format")
+    }
+}
+
+impl Into<String> for PaymentArgs {
+    fn into(self) -> String {
+        serde_json::to_string(&self).unwrap()
+    }
+}
+
+/**
+ * Fungible token-related declarations
+ */
+
+// Interface of fungible tokens
+#[near_sdk::ext_contract(ft_contract)]
+trait FungibleTokenContract {
+    fn ft_transfer(receiver_id: String, amount: String, memo: Option<String>);
+}
+
+///
+/// This contract
+#[near_bindgen]
+#[derive(Default, BorshDeserialize, BorshSerialize)]
+pub struct FungibleProxy {}
+
+// Callback methods
+#[near_sdk::ext_contract(ext_self)]
+pub trait ExtSelfRequestProxy {
+    fn on_transfer_with_reference(
+        &self,
+        args: PaymentArgs,
+        token_address: AccountId,
+        payer: AccountId,
+        amount: U128,
+    ) -> String;
+}
+
+trait FungibleTokenReceiver {
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: String, msg: String) -> Promise;
+}
+
+#[near_bindgen]
+impl FungibleTokenReceiver for FungibleProxy {
+    /// This is the function that will be called by the fungible token contract's `ft_transfer_call` function.
+    /// We use the `msg` field to obtain the arguments supplied by the caller specifying the intended payment.
+    /// `msg` should be a string in JSON format containing all the fields in `PaymentArgs`.
+    /// Eg. msg = {"payment_reference":"abc7c8bb1234fd12","to":"dummy.payee.near","fee_address":"fee.requestfinance.near","fee_amount":"200"}
+    ///
+    /// For more information on the fungible token standard, see https://nomicon.io/Standards/Tokens/FungibleToken/Core
+    ///
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: String, msg: String) -> Promise {
```

Comment:
Yes but it does not need to be `pub`, probably because it's called from another contract through shards, it's not an import. Example in the `near-sdk-rs` repo: https://github.com/near/near-sdk-rs/blob/master/examples/fungible-token/test-contract-defi/src/lib.rs#L41